test/integration/etcd_storage_path_test.go

```diff
@@ -0,0 +1,775 @@
+	reflect.TypeOf(&apisextensionsv1beta1.DeploymentRollback{}):         {ephemeral: true}, // TODO(mo): Just making the test pass
+	reflect.TypeOf(&apisextensionsv1beta1.DaemonSet{}):                  {ephemeral: true}, // TODO(mo): Just making the test pass
+	reflect.TypeOf(&apisextensionsv1beta1.Deployment{}):                 {ephemeral: true}, // TODO(mo): Just making the test pass
+	reflect.TypeOf(&apisextensionsv1beta1.ReplicationControllerDummy{}): {ephemeral: true}, // TODO(mo): Just making the test pass
+	reflect.TypeOf(&apisextensionsv1beta1.ThirdPartyResource{}):         {ephemeral: true}, // TODO(mo): Just making the test pass
+}
+
+// namespace used for all tests, do not change this
+const testNamespace = "etcdstoragepathtestnamespace"
+
+// TestEtcdStoragePath tests to make sure that all objects are stored in an expected location in etcd.
+// It will start failing when a new type is added to ensure that all future types are added to this test.
+// It will also fail when a type gets moved to a different location. Be very careful in this situation because
+// it essentially means that you will be break old clusters unless you create some migration path for the old data.
+func TestEtcdStoragePath(t *testing.T) {
+	etcdServer := testutil.RequireEtcd(t)
+	defer testutil.DumpEtcdOnFailure(t)
+	keys := etcd.NewKeysAPI(etcdServer.Client)
+
+	masterConfig, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("error getting master config: %#v", err)
+	}
+	masterConfig.AdmissionConfig.PluginOrderOverride = []string{"PodNodeSelector"} // remove most admission checks to make testing easier
+
+	kubeConfigFile, err := testserver.StartConfiguredMaster(masterConfig)
+	if err != nil {
+		t.Fatalf("error starting server: %#v", err)
+	}
+	kubeClient, err := testutil.GetClusterAdminKubeClient(kubeConfigFile)
+	if err != nil {
+		t.Fatalf("error getting client: %#v", err)
+	}
+
+	loader := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeConfigFile}, &clientcmd.ConfigOverrides{})
+	f := osclientcmd.NewFactory(loader)
+	d := f.Decoder(false)
+	seen := map[reflect.Type]struct{}{}
+
+	if _, err := kubeClient.Core().Namespaces().Create(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: testNamespace}}); err != nil {
+		t.Fatalf("error creating test namespace: %#v", err)
+	}
+
+	for _, gv := range registered.RegisteredGroupVersions() {
+		for kind, apiType := range kapi.Scheme.KnownTypes(gv) {
+			// we don't care about lists // TODO make sure this is always true
+			if strings.HasSuffix(apiType.Name(), "List") {
+				continue
+			}
+
+			ptrType := reflect.PtrTo(apiType)
+			seen[ptrType] = struct{}{}
+			testData, ok := etcdStorageData[ptrType]
+			pkgPath := apiType.PkgPath()
+
+			if !ok {
+				t.Errorf("no test data for %s from %s", kind, pkgPath)
+				continue
+			}
+
+			if testData.ephemeral {
```

Comment:
This worries me a bit wrt. to newly added resources. It might be overused by some not to write a proper test. Can we get that information from storage, that a particular resource should not be stored?